File: src/api/reader.ts
/**
 * Reader — the read API. RESOLVE / READ / LIST / SEARCH / GET_PROVENANCE
 * and READ_STREAM on a pinned commit (or an append stream). Every object
 * result carries repository/commit/object coordinates.
 *
 * Target may be a KnowledgeRef (assembled entity) or a KnowledgeAddress
 * (one aspect/member). AspectSelector is a read/index strategy, not a write.
 */

import type {
  AspectSelector,
  CommitIdentity,
  KnowledgeAddress,
  KnowledgeRef,
  KnowledgeValue,
  ObjectDiff,
  ObjectIdentity,
  ObjectRevision,
  ProvenanceTrace,
  Resolution,
  Repository,
  StreamSlice,
} from "../contracts/index.ts";
import { selectAspects } from "../contracts/access.ts";
import { IngressError } from "../contracts/errors.ts";
import type { Store } from "../store.ts";

export class Reader {
  constructor(private readonly store: Store) {}

  private repo(ref: KnowledgeRef): Repository {
    return this.repoById(ref.repository);
  }

  private repoById(repositoryId: string): Repository {
    const repo = this.store.repos.get(repositoryId);
    if (!repo) throw new IngressError("KNOWLEDGE_REF_UNRESOLVED", `unknown repository ${repositoryId}`);
    return repo;
  }

  resolve(ref: KnowledgeRef, commitId: CommitIdentity): Resolution {
    return this.repo(ref).resolve(ref.object, commitId);
  }

  read(ref: KnowledgeRef, commitId: CommitIdentity, selector?: AspectSelector): KnowledgeValue {
    const value = this.repo(ref).read(ref.object, commitId);
    if (!selector) return value;
    return { ...value, value: selectAspects(value.value, value.units, selector) };
  }

  resolveAddress(repositoryId: string, address: KnowledgeAddress, commitId: CommitIdentity): Resolution {
    return this.repoById(repositoryId).resolveAddress(address, commitId);
  }

  readAddress(repositoryId: string, address: KnowledgeAddress, commitId: CommitIdentity): KnowledgeValue {
    return this.repoById(repositoryId).readAddress(address, commitId);
  }

  /** GET_PROVENANCE: envelopes on this object's units. Does not walk refs or git log. */
  getProvenance(ref: KnowledgeRef, commitId: CommitIdentity): ProvenanceTrace {
    return this.repo(ref).getProvenance(ref.object, commitId);
  }

  readStream(repositoryId: string, streamRef: string): StreamSlice {
    return this.repoById(repositoryId).readStream(streamRef);
  }

  search(query: string, repositoryId: string, commitId: CommitIdentity): KnowledgeValue[] {
    const repo = this.store.repos.get(repositoryId);
    if (!repo) throw new IngressError("KNOWLEDGE_REF_UNRESOLVED", `unknown repository ${repositoryId}`);
    return repo.search(query, commitId);
  }

  list(repositoryId: string, commitId: CommitIdentity): KnowledgeValue[] {
    const repo = this.store.repos.get(repositoryId);
    if (!repo) throw new IngressError("KNOWLEDGE_REF_UNRESOLVED", `unknown repository ${repositoryId}`);
    return repo.list(commitId);
  }

  log(repositoryId: string, objectId: ObjectIdentity, commitId: CommitIdentity, limit?: number): readonly ObjectRevision[] {
    return this.repoById(repositoryId).log(objectId, commitId, limit);
  }

  diff(
    repositoryId: string,
    objectId: ObjectIdentity,
    fromCommit: CommitIdentity,
    toCommit: CommitIdentity,
  ): ObjectDiff {
    return this.repoById(repositoryId).diff(objectId, fromCommit, toCommit);
  }
}

File: reader/reader.go
package reader

import (
	"kc/internal/journal"
	"kc/kernel"
	"kc/repository"
)

// Reader is the read face on a pinned Repository commit. It does not write,
// does not assemble a full ViewReadVersion, and does not own Catalog federation.
//
// Tasks, by what they answer (design ch.7):
//
//	identity:  Resolve, Read, ResolveAddress, ReadAddress, List
//	history:   Log, Diff, GetProvenance
//	schema:    DescribeSchema  (schema/* AccessHints; not a GraphQL runtime)
//	stream:    QueryStream, ReadStream
//	debug:     Search  (JSON contains; production retrieval is Projection)
//
// GroundingCitation is a consume-side projection of a READ result (D12).
// Refine is an optional Ref-preserving capability, not a base read.
type Reader struct {
	store   *repository.Store
	journal journal.Journal
}

func NewReader(store *repository.Store) *Reader {
	return &Reader{store: store}
}

func (r *Reader) SetJournal(j journal.Journal) { r.journal = j }

func (r *Reader) note(cmd string, refs map[string]any, err error) error {
	return journal.Finish(r.journal, journal.LayerSystem, "reader", cmd, refs, err)
}

func (r *Reader) repoByID(repositoryID kernel.RepositoryID) (repository.Repository, error) {
	return r.store.Require(repositoryID, kernel.ErrKnowledgeRefUnresolved)
}

func (r *Reader) Resolve(ref kernel.KnowledgeRef, commitID kernel.CommitID) (resolution repository.Resolution, err error) {
	defer func() {
		err = r.note("resolve", map[string]any{"repositoryId": string(ref.Repository), "object": string(ref.Object), "commit": string(commitID)}, err)
	}()
	repo, err := r.repoByID(ref.Repository)
	if err != nil {
		return repository.Resolution{}, err
	}
	return repo.Resolve(ref.Object, commitID)
}

func (r *Reader) Read(ref kernel.KnowledgeRef, commitID kernel.CommitID, selector *repository.AspectSelector) (value repository.KnowledgeValue, err error) {
	defer func() {
		err = r.note("read", map[string]any{"repositoryId": string(ref.Repository), "object": string(ref.Object), "commit": string(commitID)}, err)
	}()
	repo, err := r.repoByID(ref.Repository)
	if err != nil {
		return repository.KnowledgeValue{}, err
	}
	value, err = repo.Read(ref.Object, commitID)
	if err != nil {
		return repository.KnowledgeValue{}, err
	}
	if selector == nil {
		return value, nil
	}
	value.Value = repository.SelectAspects(value.Value, value.Units, selector)
	return value, nil
}

func (r *Reader) ResolveAddress(repositoryID kernel.RepositoryID, address kernel.Address, commitID kernel.CommitID) (resolution repository.Resolution, err error) {
	defer func() {
		err = r.note("resolve-address", map[string]any{"repositoryId": string(repositoryID), "object": string(address.ObjectID), "commit": string(commitID)}, err)
	}()
	repo, err := r.repoByID(repositoryID)
	if err != nil {
		return repository.Resolution{}, err
	}
	return repo.ResolveAddress(address, commitID)
}

func (r *Reader) ReadAddress(repositoryID kernel.RepositoryID, address kernel.Address, commitID kernel.CommitID) (value repository.KnowledgeValue, err error) {
	defer func() {
		err = r.note("read-address", map[string]any{"repositoryId": string(repositoryID), "object": string(address.ObjectID), "commit": string(commitID)}, err)
	}()
	repo, err := r.repoByID(repositoryID)
	if err != nil {
		return repository.KnowledgeValue{}, err
	}
	return repo.ReadAddress(address, commitID)
}

func (r *Reader) List(repositoryID kernel.RepositoryID, commitID kernel.CommitID) (values []repository.KnowledgeValue, err error) {
	defer func() {
		err = r.note("list", map[string]any{"repositoryId": string(repositoryID), "commit": string(commitID)}, err)
	}()
	repo, err := r.repoByID(repositoryID)
	if err != nil {
		return nil, err
	}
	return repo.List(commitID)
}
